engine/common/version/version_control.go

```diff
@@ -0,0 +1,314 @@
+func (v *VersionControl) CompatibleAtBlock(height uint64) (bool, error) {
+	// Check if the height is greater than the last handled block height. If so, return an error indicating that the height is unhandled.
+	if height > v.lastProcessedHeight.Load() {
+		return false, fmt.Errorf("could not check compatibility for height %d: last handled height is %d: %w", height, v.lastProcessedHeight.Load(), ErrOutOfRange)
+	}
+
+	startHeight := v.startHeight.Load()
+	// Check if the start height is set and the height is less than the start height. If so, return false indicating that the height is not compatible.
+	if startHeight != NoHeight && height < startHeight {
+		return false, nil
+	}
+
+	endHeight := v.endHeight.Load()
+	// Check if the end height is set and the height is greater than the end height. If so, return false indicating that the height is not compatible.
+	if endHeight != NoHeight && height > endHeight {
+		return false, nil
+	}
+
+	// If none of the above conditions are met, the height is compatible.
+	return true, nil
+}
+
+// AddVersionUpdatesConsumer adds a consumer for version update events.
+func (v *VersionControl) AddVersionUpdatesConsumer(consumer VersionControlConsumer) {
+	v.Lock()
+	defer v.Unlock()
+
+	v.consumers = append(v.consumers, consumer)
+}
+
+// processEvents is a worker that processes block finalized events.
+func (v *VersionControl) processEvents(
+	ctx irrecoverable.SignalerContext,
+	ready component.ReadyFunc,
+) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-v.finalizedHeightNotifier.Channel():
+			v.blockFinalized(ctx, v.finalizedHeight.Value())
+		}
+	}
+}
+
+// blockFinalized processes a block finalized event and updates the version control state.
+func (v *VersionControl) blockFinalized(
+	ctx irrecoverable.SignalerContext,
+	newFinalizedHeight uint64,
+) {
+	lastProcessedHeight := v.lastProcessedHeight.Load()
+	if lastProcessedHeight >= newFinalizedHeight {
+		// already processed this or a higher version beacon
+		return
+	}
+
+	for height := lastProcessedHeight + 1; height <= newFinalizedHeight; height++ {
+		vb, err := v.versionBeacons.Highest(height)
+		if err != nil {
+			v.log.Err(err).
+				Uint64("height", height).
+				Msg("Failed to get highest version beacon for version control")
+
+			ctx.Throw(
+				fmt.Errorf(
+					"failed to get highest version beacon for version control: %w",
+					err))
+			return
+		}
+
+		if vb == nil {
+			// no version beacon found
+			// this is unexpected as there should always be at least the
+			// starting version beacon, but not fatal.
+			// It can happen if the node starts before bootstrap is finished.
+			// TODO: remove when we can guarantee that there will always be a version beacon
+			v.log.Info().
+				Uint64("height", height).
+				Msg("No version beacon found for version control")
+			continue
+		}
+
+		v.lastProcessedHeight.Store(height)
+
+		// version boundaries are sorted by version
```

Comment:
source: https://github.com/onflow/flow-core-contracts/blob/master/contracts/NodeVersionBeacon.cdc#L198-L205

```suggestion
		// version boundaries are sorted by blockHeight in ascending order
		// the first version greater than the node's is the version transition height
```